File: problems/chapter06/miiitomi/005.cpp
// AtCoder ARC037   C - 億マス計算
// https://atcoder.jp/contests/arc037/tasks/arc037_c

#include <bits/stdc++.h>
using namespace std;

int N, K;
vector<int> a, b;

bool f(int64_t m) {
    int counter = 0;
    for (int i = 0; i < N; i++) {
        int64_t x = (m / a[i]) + 1;
        auto iter = lower_bound(b.begin(), b.end(), x);
        counter += distance(b.begin(), iter);
        if (counter >= K) return true;
    }
    return false;
}

int main() {
    cin >> N >> K;
    a.resize(N);
    b.resize(N);
    for (int i = 0; i < N; i++) cin >> a[i];
    for (int i = 0; i < N; i++) cin >> b[i];
    sort(a.begin(), a.end());
    sort(b.begin(), b.end());

    int64_t left = 0;
    int64_t right = (int64_t)a[N-1] * b[N-1];
    int64_t mid;

    while (right - left > 1) {
        mid = (left + right) / 2;
        if (f(mid)) {
            right = mid;
        } else {
            left = mid;
        }
    }
    cout << right << endl;
}

File: problems/chapter06/mshibatatt/005.py
# https://atcoder.jp/contests/arc037/tasks/arc037_c

import bisect

def less_than_k(x, a, b, k):
    # fix a and binary search in b
    # count number of a * b < x
    # return True if count < k otherwise False
    counter = 0
    for i in a:
        counter += bisect.bisect_right(b, x//i)
        if counter >= k:
            return True
    return False

def main():
    N, k = map(int, input().split())
    a = list(map(int, input().split()))
    b = list(map(int, input().split()))

    a.sort()
    b.sort()

    left = 0
    right = a[-1]*b[-1]+1

    while right - left > 1:
        mid = left + (right - left)//2
        if less_than_k(mid, a, b, k):
            right = mid
        else:
            left = mid

    print(right)

if __name__=='__main__':
    main()